[SPARK-29224][ML]Implement Factorization Machines as a ml-pipeline component

mob-ai opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
-->

### What changes were proposed in this pull request?

Implement Factorization Machines as a ml-pipeline component

1. loss function supports: logloss, mse
2. optimizer: GD, adamW


### Why are the changes needed?

Factorization Machines is widely used in advertising and recommendation system to estimate CTR(click-through rate).
Advertising and recommendation system usually has a lot of data, so we need Spark to estimate the CTR, and Factorization Machines are common ml model to estimate CTR.
References:

1. S. Rendle, “Factorization machines,” in Proceedings of IEEE International Conference on Data Mining (ICDM), pp. 995–1000, 2010.
https://www.csie.ntu.edu.tw/~b97053/paper/Rendle2010FM.pdf


### Does this PR introduce any user-facing change?

No

### How was this patch tested?

run unit tests

mobai-zhanjf committed:
add Factorization Machines

mobai-zhanjf committed:
fix bug: Param[Boolean] change to BooleanParam

mobai-zhanjf committed:
update solver and loss api

mobai-zhanjf committed:
update alias oldXXX to OldXXX

mobai-zhanjf committed:
update Since2.4.3 to Since3.0.0

mobai-zhanjf committed:
add more unit test

mobai-zhanjf committed:
impl python FM

mobai-zhanjf committed:
update coeff init

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

mob-ai commented:
Jenkins, retest this please

mob-ai commented:
Jenkins, retest this please

mob-ai commented:
retest this please

mobai-zhanjf committed:
update comments

mobai-zhanjf committed:
update fm python style

mobai-zhanjf committed:
fix python style

mob-ai commented:
@srowen I don't know how to change PR target branch, so I open a new PR. I have checkout branch base master(spark 3.0.0), and implement FM python api.
@zhengruifeng I plat the convergence curves use normal GD and adamW. https://issues.apache.org/jira/browse/SPARK-29224

srowen commented on mllib/src/main/scala/org/apache/spark/ml/regression/FactorizationMachines.scala:
```diff
@@ -0,0 +1,839 @@
+import org.apache.spark.storage.StorageLevel
+
+/**
+ * Params for Factorization Machines
```
Is there any paper you can link to to explain the implementation? or, just a few paragraphs about what the implementation does?

mob-ai replied:
FM paper: S. Rendle, “Factorization machines,” in Proceedings of IEEE International Conference on Data Mining (ICDM), pp. 995–1000, 2010.
https://www.csie.ntu.edu.tw/~b97053/paper/Rendle2010FM.pdf

mob-ai replied:
> Is there any paper you can link to to explain the implementation? or, just a few paragraphs about what the implementation does?

FM paper: S. Rendle, “Factorization machines,” in Proceedings of IEEE International Conference on Data Mining (ICDM), pp. 995–1000, 2010.
https://www.csie.ntu.edu.tw/~b97053/paper/Rendle2010FM.pdf

srowen replied:
Yes, I mean, this should be in the docs along with a brief explanation of what it does ... enough that an informed reader understands how the parameters you expose map to the algorithm in the paper (which is probably straightforward)

srowen commented on mllib/src/main/scala/org/apache/spark/ml/regression/FactorizationMachines.scala:
```diff
@@ -0,0 +1,839 @@
+   */
+  @Since("3.0.0")
+  final val numFactors: IntParam = new IntParam(this, "numFactors",
+    "dimensionality of the factorization")
```
These doc comments don't need to be long, but I think it could explain a little bit more. I think this text is OK if the 'dimension' in question is clear from an explanation in the class doc, or a paper.

srowen commented on mllib/src/main/scala/org/apache/spark/ml/regression/FactorizationMachines.scala:
```diff
@@ -0,0 +1,839 @@
+   */
+  @Since("3.0.0")
+  final val regParam: DoubleParam = new DoubleParam(this, "regParam",
+    "regularization for L2")
```
This could be clarified a bit: strength of regularization? does this correspond to some parameter in a paper, some 'alpha'? etc . You can add a validator to ensure it's not negative, too. See some other DoubleParam for an example.

srowen commented on mllib/src/main/scala/org/apache/spark/ml/regression/FactorizationMachines.scala:
```diff
@@ -0,0 +1,839 @@
+   * @group param
+   */
+  @Since("3.0.0")
+  final val miniBatchFraction: DoubleParam = new DoubleParam(this, "miniBatchFraction",
```
Likewise, minibatch of what? and can be validated to be nonnegative/positive?

srowen commented on mllib/src/main/scala/org/apache/spark/ml/regression/FactorizationMachines.scala:
```diff
@@ -0,0 +1,839 @@
+   * @group param
+   */
+  @Since("3.0.0")
+  final val verbose: BooleanParam = new BooleanParam(this, "verbose",
```
I wouldn't add this; we don't elsewhere. Just use logging levels appropriately.

srowen commented on mllib/src/main/scala/org/apache/spark/ml/regression/FactorizationMachines.scala:
```diff
@@ -0,0 +1,839 @@
+        (if ($(fitLinear)) Array.fill(numFeatures)(0.0) else Array.empty[Double]) ++
+        (if ($(fitBias)) Array.fill(1)(0.0) else Array.empty[Double]))
+
+    val data = instances.map{ case OldLabeledPoint(label, features) => (label, features) }
```
Nit: space before brace

srowen commented on mllib/src/main/scala/org/apache/spark/ml/regression/FactorizationMachines.scala:
```diff
@@ -0,0 +1,839 @@
+    }
+    (0 until numFactors).foreach { f =>
+      var sumSquare = 0.0
+      var squareSum = 0.0
```
Nit: this looks like a 'sum' not squared sum.

srowen commented on mllib/src/main/scala/org/apache/spark/ml/regression/FactorizationMachines.scala:
```diff
@@ -0,0 +1,839 @@
+}
+
+private[ml] class AdamWUpdater(weightSize: Int) extends Updater with Logging {
+  var beta1: Double = 0.9
```
Can these be val?

srowen commented on python/pyspark/ml/regression.py:
```diff
@@ -2126,6 +2127,181 @@ def __repr__(self):
+
+    loss Supports:
+
+    * logisticLoss (default)
```
Dumb question but does this render correctly in python docs? I'm not sure this is how bulleted lists work, but am not sure

mob-ai replied:
> Dumb question but does this render correctly in python docs? I'm not sure this is how bulleted lists work, but am not sure

I refer to the LinearRegression docs (regression.py: line51). In my experience, python __doc__ just a string, it won't render as a list. But I have not idea, whether spark will use __doc__ to generate docs html.

srowen reviewed (changes requested):
Looks promising

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/regression/FactorizationMachines.scala:
```diff
@@ -0,0 +1,839 @@
+  @Since("3.0.0")
+  final override val loss: Param[String] = new Param[String](this, "loss", "The loss function to" +
+    s" be optimized. Supported options: ${supportedLosses.mkString(", ")}. (Default logisticLoss)",
+    ParamValidators.inArray[String](supportedLosses))
```
for a log-loss, should we put it into `ml.classification`?

mob-ai replied:
> for a log-loss, should we put it into `ml.classification`?

I will add a FM Classifier later.

srowen replied:
I guess the question is then, do you want to add log-loss later? it won't make sense for a regression

mob-ai replied:
@srowen When loss is logloss, labels must in {0, 1}. So, I am planning to add FMClassifier for logloss, original FactorizationMachines changes to FMRegressor for mse, then remove loss parameter. What do you think?

srowen replied:
I would implement it all in one pull request then. My concern is that we're coming up on Spark 3 and we would not want to release part of this only to change the API after.

mob-ai replied:
OK, I will implement Classifier and Regressor in this PR. Thank you for your suggest.

mobai-zhanjf committed:
resolve change requested

srowen commented on mllib/src/main/scala/org/apache/spark/ml/regression/FactorizationMachines.scala:
```diff
@@ -0,0 +1,757 @@
+  import FactorizationMachines._
+
+  /**
+   * Param for dimensionality of the factors (&gt;= 0)
```
Nit: Must be greater than 0 right? the check below is correct

srowen commented on mllib/src/main/scala/org/apache/spark/ml/regression/FactorizationMachines.scala:
```diff
@@ -0,0 +1,757 @@
+   */
+  @Since("3.0.0")
+  final val numFactors: IntParam = new IntParam(this, "numFactors",
+    "dimensionality of the factor vectors, " +
```
Nit: the doc strings should start with a Capital

srowen commented on mllib/src/main/scala/org/apache/spark/ml/regression/FactorizationMachines.scala:
```diff
@@ -0,0 +1,757 @@
+   */
+  @Since("3.0.0")
+  final val regParam: DoubleParam = new DoubleParam(this, "regParam",
+    "the parameter of l2-regularization term, " +
```
Nit: l2 -> L2
Nit: can we say "strength" rather than "parameter" to clarify a little?
Rather than describe what L2 regularization is, what is being regularized? i'm just looking for a tiny bit more specificity.

srowen commented on mllib/src/main/scala/org/apache/spark/ml/regression/FactorizationMachines.scala:
```diff
@@ -0,0 +1,757 @@
+   */
+  @Since("3.0.0")
+  def setLoss(value: String): this.type = set(loss, value)
+  setDefault(loss -> LogisticLoss)
```
Maybe I'm just not familiar here, but if this is a regressor, can you even use logistic loss?

srowen commented on mllib/src/main/scala/org/apache/spark/ml/regression/FactorizationMachines.scala:
```diff
@@ -0,0 +1,757 @@
+      (if ($(fitBias)) 1 else 0)
+    val initialCoefficients =
+      Vectors.dense(Array.fill($(numFactors) * numFeatures)(Random.nextGaussian() * $(initStd)) ++
+        (if ($(fitLinear)) Array.fill(numFeatures)(0.0) else Array.empty[Double]) ++
```
Nit: where you call `Array.fill(...)(0.0)` just `new Array[Double](...)`. It doesn't matter much but no need to set the array to 0 as this is already its value. I dont' feel strongly about it.

zhengruifeng replied:
another nit, `Array.emptyDoubleArray`

mob-ai replied:
> another nit, `Array.emptyDoubleArray`

The difference between `Array.emptyDoubleArray` and `Array.empty[Double]`?

mob-ai replied:
> another nit, `Array.emptyDoubleArray`

`Array.emptyDoubleArray` is a val, `Array.empty[Double]` will create a new object. So `Array.emptyDoubleArray` is more efficient, right?

srowen commented on python/pyspark/ml/regression.py:
```diff
@@ -2126,6 +2127,172 @@ def __repr__(self):
+    """
+
+    numFactors = Param(Params._dummy(), "numFactors", "dimensionality of the factor vectors, " +
+                       "which are used to get pairwise interactions between variables",
```
Just a note to update the strings here afterwards to match Scala

srowen replied:
Heh, I've lost track of what I was referring to here. Is it resolved? :)

mob-ai replied:
It is resolved. The doc strings should start with a Capital (dimensionality of .. -> Dimensionality of ...).

mobai-zhanjf committed:
resolve change requests

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/regression/FactorizationMachines.scala:
```diff
@@ -0,0 +1,757 @@
+    train(dataset, handlePersistence)
+  }
+
+  protected[spark] def train(
```
I guess we do not need to define this function. Other impls have this function for call in the `.mllib` side.

mob-ai replied:
> I guess we do not need to define this function. Other impls have this function for call in the `.mllib` side.

Yes, ml's Estimator use `.fit()` public function (not `.train()`). But in `.fit() ` source code, it will do something then call a train function, so I need to implement it. The train function is not a public function, users can only use fit function.

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/regression/FactorizationMachines.scala:
```diff
@@ -0,0 +1,757 @@
+      case AdamW => new AdamWUpdater(coefficientsSize)
+    }
+
+    val optimizer = new GradientDescent(gradient, updater)
```
BTW, If both FM and MLP still need mini-batch solver, we may move it to the .ml side in the future, to avoid vector conversion.

mob-ai replied:
> BTW, If both FM and MLP still need mini-batch solver, we may move it to the .ml side in the future, to avoid vector conversion.

I agree, and I think GradientDescent supports aggregateDepth parameter, seed parameter, weightCol parameter and logInfo loss value per epoch will be better.

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/regression/FactorizationMachines.scala:
```diff
@@ -0,0 +1,757 @@
+@Since("3.0.0")
+class FactorizationMachinesModel (
+    @Since("3.0.0") override val uid: String,
+    @Since("3.0.0") val coefficients: OldVector,
```
I'd prefer a `ml.vector` here, just call `.asML` method.

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/regression/FactorizationMachines.scala:
```diff
@@ -0,0 +1,757 @@
+
+  protected def getLoss(rawPrediction: Double, label: Double): Double
+
+  private val sumVX = Array.fill(numFactors)(0.0)
```
will this cause potential thread safety issue?

mob-ai replied:
> will this cause potential thread safety issue?

I think it is thread safety.

The GD calculates the gradient code: (spark/mllib/optimization/GradientDescent.scala: 239)
GD will use treeAggregate to calculate gradient.

```scala
        .treeAggregate((BDV.zeros[Double](n), 0.0, 0L))(
          seqOp = (c, v) => {
            // c: (grad, loss, count), v: (label, features)
            val l = gradient.compute(v._2, v._1, bcWeights.value, Vectors.fromBreeze(c._1))
            (c._1, c._2 + l, c._3 + 1)
          },
          combOp = (c1, c2) => {
            // c: (grad, loss, count)
            (c1._1 += c2._1, c1._2 + c2._2, c1._3 + c2._3)
          })
```

treeAggregate code: (spark/rdd/RDD.scala: 1201)
treeAggregate will call `mapPartitions` to calculate local gradient pre partition. And in a partition, it will run `it.aggregate`.

```scala
      val cleanSeqOp = context.clean(seqOp)
      val cleanCombOp = context.clean(combOp)
      val aggregatePartition =
        (it: Iterator[T]) => it.aggregate(zeroValue)(cleanSeqOp, cleanCombOp)
      var partiallyAggregated: RDD[U] = mapPartitions(it => Iterator(aggregatePartition(it)))
```

scala aggregate code: (scala/collection/TraversableOnce.scala: 214)
`aggregate` only use seqop, I think it means cumulative gradient will be calculated sequentially in a partition.

```scala
  def aggregate[B](z: =>B)(seqop: (B, A) => B, combop: (B, B) => B): B = foldLeft(z)(seqop)
```

zhengruifeng replied:
In this PR, method `getRawPrediction` is not only used in training, but also used in prediction. What if user call method `predict` concurrently?

mob-ai replied:
> In this PR, method `getRawPrediction` is not only used in training, but also used in prediction. What if user call method `predict` concurrently?

When user call method `predict`, val `sumVX` will not be used. (`sumVX` is only used when calculate gradient)

mob-ai replied:
But I will change it to avoid more worry.

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/regression/FactorizationMachines.scala:
```diff
@@ -0,0 +1,757 @@
+}
+
+private[ml] class AdamWUpdater(weightSize: Int) extends Updater with Logging {
+  val beta1: Double = 0.9
```
these params are not exposed to end users?

mob-ai replied:
> these params are not exposed to end users?

I also considered this. But I was warried about that if someone adds better solver, then the solver's parameters will be too much. And in AdamW, these parameters almost not be tuning. So I am not exposed these parameters for now.

zhengruifeng reviewed (commented):
In general, I think it maybe better to move the impl to `ml.regression` as a private class, and add wrappers in both `ml.regression` & `ml.classification`.
And I think we need more suites, (such as withBias/withoutBias/withLinear/withoutLinear). since it will be an important impl.

zhengruifeng commented:
In practice I am using FM/FFM, and IMHO SSP or ASYNC solvers (like Difacto/PS-lite) seems more efficient than mini-batch solvers.
I'm wondering whether it is possible to include a simple PS framework(like [glint](https://github.com/MLnick/glint-fm) ) in the future.

mobai-zhanjf committed:
FM split to FMClassifier and FMRegressor

mob-ai commented:
I already resolved change requested. And I splited FM to FMClassifier and FMRegressor.

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/classification/FMClassifier.scala:
```diff
@@ -0,0 +1,360 @@
+  extends ProbabilisticClassifier[Vector, FMClassifier, FMClassifierModel]
+  with FMClassifierParams with DefaultParamsWritable with Logging {
+
+  import org.apache.spark.ml.regression.BaseFactorizationMachinesGradient.{LogisticLoss, parseLoss}
```
Can we put those import above?

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/classification/FMClassifier.scala:
```diff
@@ -0,0 +1,360 @@
+    train(dataset, handlePersistence)
+  }
+
+  protected[spark] def train(
```
I still think that this method is not needed, just create a singe `def train(dataset: Dataset[_]): FMClassifierModel`?

mob-ai replied:
> I still think that this method is not needed, just create a singe `def train(dataset: Dataset[_]): FMClassifierModel`?

@zhengruifeng I refer to the implementation of LR. It will handle persistence. If dataset not be persisted, it would persist the dataset, and release the cache after train finish.
Or should I cache dataset regardless of whether the dataset be cached?

LogisticRegression.scala: 481

```scala
  override protected[spark] def train(dataset: Dataset[_]): LogisticRegressionModel = {
    val handlePersistence = dataset.storageLevel == StorageLevel.NONE
    train(dataset, handlePersistence)
  }

  protected[spark] def train(
      dataset: Dataset[_],
      handlePersistence: Boolean): LogisticRegressionModel = instrumented { instr =>
    val instances = extractInstances(dataset)

    if (handlePersistence) instances.persist(StorageLevel.MEMORY_AND_DISK)

    // train model code

    if (handlePersistence) instances.unpersist()
```

zhengruifeng replied:
Yes, but this method `def train(dataset: Dataset[_], handlePersistence: Boolean)` is added for the old [`mllib.LogisticRegression`](https://github.com/apache/spark/blob/6378d4bc06cd1bb1a209bd5fb63d10ef52d75eb4/mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala#L345).

As a new alg, FM do not need to have this method.

>  Or should I cache dataset regardless of whether the dataset be cached?
you should cache the dataset, refer to [LinearSVC](https://github.com/apache/spark/blob/8a4378c6f0ca49f94fdeba08ca095721427e8635/mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala#L165)

zhengruifeng replied:
```scala
  override protected[spark] def train(dataset: Dataset[_]): FMClassifierModel = {
    val handlePersistence = dataset.storageLevel == StorageLevel.NONE
    val data: RDD[(Double, OldVector)] = ...
    if (handlePersistence) data.persist(StorageLevel.MEMORY_AND_DISK)
    ...
  }
```

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/classification/FMClassifier.scala:
```diff
@@ -0,0 +1,360 @@
+      .setRegParam($(regParam))
+      .setMiniBatchFraction($(miniBatchFraction))
+      .setConvergenceTol($(tol))
+    val coefficients = optimizer.optimize(data, initialCoefficients)
```
If the real training dataset is `data` not `instances`, why not cache `data` instead?

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/classification/FMClassifier.scala:
```diff
@@ -0,0 +1,360 @@
+  extends ProbabilisticClassificationModel[Vector, FMClassifierModel]
+    with FMClassifierParams with MLWritable {
+
+  import org.apache.spark.ml.regression.BaseFactorizationMachinesGradient.LogisticLoss
```
I guess we can put this import above

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/classification/FMClassifier.scala:
```diff
@@ -0,0 +1,360 @@
+   *   [i * numFactors + f] denotes i-th feature and f-th factor
+   * Following indices are 1-way coefficients and global bias.
+   */
+  private val oldCoefficients: OldVector = coefficients
```
I guess we can mark it `lazy` and `transient`?

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/classification/FMClassifier.scala:
```diff
@@ -0,0 +1,360 @@
+   */
+  private val oldCoefficients: OldVector = coefficients
+
+  private lazy val gradient = BaseFactorizationMachinesGradient.parseLoss(
```
`transient`?

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/classification/FMClassifier.scala:
```diff
@@ -0,0 +1,360 @@
+    private case class Data(
+      numFeatures: Int,
+      numClasses: Int,
+      coefficients: OldVector)
```
`coefficients` is now a `ml.Vector`, right?

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/classification/FMClassifier.scala:
```diff
@@ -0,0 +1,360 @@
+@Since("3.0.0")
+class FMClassifierModel (
+  @Since("3.0.0") override val uid: String,
+  @Since("3.0.0") val coefficients: Vector,
```
can we split `coefficients` into three part: bias/linear/factors ?
I think it is useful to obtain the factors matix in some cases.

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/regression/FMRegressor.scala:
```diff
@@ -0,0 +1,768 @@
+  final def getInitStd: Double = $(initStd)
+
+  /** String name for "gd". */
+  private[ml] val GD = "gd"
```
I think `GD`, `AdamW`, `supportedSolvers` should be defined in an object

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/regression/FMRegressor.scala:
```diff
@@ -0,0 +1,768 @@
+  with FMRegressorParams with DefaultParamsWritable with Logging {
+
+  import org.apache.spark.ml.regression.BaseFactorizationMachinesGradient.{SquaredError, parseLoss}
+  import org.apache.spark.ml.regression.FMRegressor.initCoefficients
```
ditto

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/regression/FMRegressor.scala:
```diff
@@ -0,0 +1,768 @@
+   */
+  @Since("3.0.0")
+  def setMiniBatchFraction(value: Double): this.type = {
+    require(value > 0 && value <= 1.0,
```
`ParamValidators.inRange(0, 1, false, true)` already checks input value

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/regression/FMRegressor.scala:
```diff
@@ -0,0 +1,768 @@
+      .setRegParam($(regParam))
+      .setMiniBatchFraction($(miniBatchFraction))
+      .setConvergenceTol($(tol))
+    val coefficients = optimizer.optimize(data, initialCoefficients)
```
ditto

zhengruifeng commented on mllib/src/test/scala/org/apache/spark/ml/classification/FMClassifierSuite.scala:
```diff
@@ -0,0 +1,143 @@
+import org.apache.spark.ml.util.TestingUtils._
+import org.apache.spark.sql.{DataFrame, Row}
+
+class FMClassifierSuite extends MLTest with DefaultReadWriteTest {
```
Can we compare the fitted coefficients with libFM?

mob-ai replied:
I want to compare with libFM, but randomly generate a binary dataset, FM can't fit the dataset perfectly. Generally, I get very small loss value, but the parameters are not exactly same every time (also in libFM).

My current test method is that for mse to generate a feature-interaction dataset, FMRegressor needs to fit this dataset perfectly, bias/linearVector needs same, every `<v_i, v_j>` needs same that calculates by factorMatrix.

However, for FMClassifier, it is difficult to generate a binary feature-interaction dataset, so for the FMClassifier, I don't verify the correctness of interaction logic (that is verify in mse), only verify the correctness of logloss logic.

zhengruifeng replied:
I still doubt whether existing testsuite is enough. 
1, I suggest to add several testsuites to check whether `FMClassifier` and `FMRegressor` can learn the intercept correctly. 
Like '// Test if we can correctly learn Y = 0.1 + 1.2*X1 - 1.3*X2 + 20*X1*X2' in https://github.com/apache/spark/pull/5591/files

2, I suggest to add some testcases with `fitBias==false` or/and `fitLinear==false`

srowen replied:
Is this resolved? I'd also love to have at least one test that compares output to some other library, even if the test can only assert some approximate match in loss, etc.

mob-ai replied:
> Is this resolved? I'd also love to have at least one test that compares output to some other library, even if the test can only assert some approximate match in loss, etc.

Now, test method is FM fit a dataset perfectly (loss closed to zero and model's weights closed to real weights), the dataset is generated by a group weight (include bias/linear/factor weights). I think it is enough.

mobai-zhanjf committed:
add FactorizationMachines class to control common code

mobai-zhanjf committed:
coefficients split to bias/linearVector/factorMatrix

mobai-zhanjf committed:
update fm testcase

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/classification/FMClassifier.scala:
```diff
@@ -0,0 +1,326 @@
+ * Model produced by [[FMClassifier]]
+ */
+@Since("3.0.0")
+class FMClassifierModel (
```
`class FMClassifierModel private[classification]`

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/classification/FMClassifier.scala:
```diff
@@ -0,0 +1,326 @@
+  @Since("3.0.0") val bias: Double,
+  @Since("3.0.0") val linearVector: Vector,
+  @Since("3.0.0") val factorMatrix: Matrix,
+  @Since("3.0.0") override val numFeatures: Int,
```
move `numFeatures`, `numClasses` outof `FMClassifierModel`
refer to [LinearSVCModel](https://github.com/apache/spark/blob/5853e8b3301fd7b0bff721d5a47139afb17bfd2b/mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala#L287)

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/classification/FMClassifier.scala:
```diff
@@ -0,0 +1,326 @@
+   * @group setParam
+   */
+  @Since("3.0.0")
+  def setNumFactors(value: Int): this.type = set(numFactors, value)
```
I guess the var name `numFactors` is misleading,  #factors should be equal to #features.

refer to the name `vectorSize` in [W2V](https://github.com/apache/spark/blob/e7443d6412582aa16769e298c31d889a5ba0143c/mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala#L47), I perfer rename `numFactors` to `factorSize`

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/classification/FMClassifier.scala:
```diff
@@ -0,0 +1,326 @@
+    instr.logNumFeatures(numFeatures)
+
+    val coefficients = _train(data, numFeatures, LogisticLoss)
+
```
`if (handlePersistence) data.unpersist()`

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/classification/FMClassifier.scala:
```diff
@@ -0,0 +1,326 @@
+    val numFeatures = data.first()._2.size
+    instr.logNumFeatures(numFeatures)
+
+    val coefficients = _train(data, numFeatures, LogisticLoss)
```
we do not name a method like `_train`, what about `trainImpl` or `FactorizationMachines.train`?

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/classification/FMClassifier.scala:
```diff
@@ -0,0 +1,326 @@
+    LogisticLoss, $(numFactors), $(fitBias), $(fitLinear), numFeatures)
+
+  override protected def predictRaw(features: Vector): Vector = {
+    val rawPrediction: Double = gradient.getRawPrediction(features, oldCoefficients)
```
I'd perfer to define another prediction method using `bias`/`weights`/`factors` directly, to avoid  conversions to `mlllib.vector`

zhengruifeng replied:
I guess then we do not need `oldCoefficients`/`gradient`

mob-ai replied:
> I guess then we do not need `oldCoefficients`/`gradient`

emm...But then I have to implement the predict code twice...Actually I want to abstract it into one class.

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/classification/FMClassifier.scala:
```diff
@@ -0,0 +1,326 @@
+class FMClassifierModel (
+  @Since("3.0.0") override val uid: String,
+  @Since("3.0.0") val bias: Double,
+  @Since("3.0.0") val linearVector: Vector,
```
what about renaming them to `linear`/`weights` & `factors`?

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/classification/FMClassifier.scala:
```diff
@@ -0,0 +1,326 @@
+    val (bias, linearVector, factorMatrix) = splitCoefficients(
+      coefficients, numFeatures, $(numFactors), $(fitBias), $(fitLinear))
+
+    val model = copyValues(new FMClassifierModel(uid,
```
`copyValues(new FMClassifierModel(uid, bias, linearVector, factorMatrix, numFeatures, numClasses))`
is enough, you do not need to define a var `model`

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/regression/FMRegressor.scala:
```diff
@@ -0,0 +1,786 @@
+   * @group param
+   */
+  @Since("3.0.0")
+  final val regParam: DoubleParam = new DoubleParam(this, "regParam",
```
why not extending `HasRegParam`?

mob-ai replied:
> why not extending `HasRegParam`?

Because I want to change `regParam` doc (FM regParams only for L2), but `HasRegParam` has `final` tag, I can't override it.

zhengruifeng replied:
this regparam is appled to all coefficients, right?

srowen replied:
Hm, I think I might still use HasRegParam here and clarify the doc perhaps in setRegParam. It's a little wonky but probably worth it to make it fit the class hierarchy

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/regression/FMRegressor.scala:
```diff
@@ -0,0 +1,786 @@
+   * @group param
+   */
+  @Since("3.0.0")
+  final override val solver: Param[String] = new Param[String](this, "solver",
```
Why not extending `HasSolver`?

mob-ai replied:
> Why not extending `HasSolver`?

I already extended `HasSolver`, and I override solver to change doc.

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/regression/FMRegressor.scala:
```diff
@@ -0,0 +1,786 @@
+    initialCoefficients
+  }
+
+  private[ml] def _train(
```
_train -> train

hahadsg commented on mllib/src/main/scala/org/apache/spark/ml/classification/FMClassifier.scala:
```diff
@@ -0,0 +1,326 @@
+            s" FMClassifier currently only supports binary classification.")
+          (label, features)
+      }
+    data.persist(StorageLevel.MEMORY_AND_DISK)
```
> ```scala
> val instances = extractInstances(dataset)
> if (handlePersistence) data.persist(StorageLevel.MEMORY_AND_DISK)
> ```

@zhengruifeng `extractInstances` function return `RDD[Instance]`, but I need `RDD[(Double, OldVector)]` in train stage. Should I still get instances?

mob-ai replied:
> ```scala
> val instances = extractInstances(dataset)
> if (handlePersistence) data.persist(StorageLevel.MEMORY_AND_DISK)
> ```

@zhengruifeng  `extractInstances` function return `RDD[Instance]`, but I need `RDD[(Double, OldVector)]` in train stage. Should I still get instances?

mob-ai replied:
> ```scala
> val instances = extractInstances(dataset)
> if (handlePersistence) data.persist(StorageLevel.MEMORY_AND_DISK)
> ```

@zhengruifeng `handlePersistence` need be implemented like this.

```
  override protected[spark] def train(dataset: Dataset[_]): LogisticRegressionModel = {
    val handlePersistence = dataset.storageLevel == StorageLevel.NONE
    train(dataset, handlePersistence)
  }

  protected[spark] def train(
      dataset: Dataset[_],
      handlePersistence: Boolean): LogisticRegressionModel = instrumented { instr =>
    val instances = extractInstances(dataset)

    if (handlePersistence) instances.persist(StorageLevel.MEMORY_AND_DISK)

    // train model code

    if (handlePersistence) instances.unpersist()
```

But in previous review, you told me that I should not implement `handlePersistence`, because It's implement for mllib LR. I am confused.

mobai-zhanjf committed:
change some details

mobai-zhanjf committed:
change sumVX logic

mobai-zhanjf committed:
add handlePersistence

mobai-zhanjf committed:
remove unused import

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/classification/FMClassifier.scala:
```diff
@@ -0,0 +1,333 @@
+ * Model produced by [[FMClassifier]]
+ */
+@Since("3.0.0")
+class FMClassifierModel private[classification] (
```
`FMClassifierModel` -> `FMClassificationModel`

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/classification/FMClassifier.scala:
```diff
@@ -0,0 +1,333 @@
+  override def load(path: String): FMClassifierModel = super.load(path)
+
+  /** [[MLWriter]] instance for [[FMClassifierModel]] */
+  private[FMClassifierModel] class FMClassifierModelWriter(
```
`FMClassifierModelWriter` -> `FMClassificationModelWriter`

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/classification/FMClassifier.scala:
```diff
@@ -0,0 +1,333 @@
+    }
+  }
+
+  private class FMClassifierModelReader extends MLReader[FMClassifierModel] {
```
`FMClassifierModelReader` -> `FMClassificationModelReader`

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/regression/FMRegressor.scala:
```diff
@@ -0,0 +1,796 @@
+  private[ml] def initCoefficients(numFeatures: Int): OldVector = {
+    val initialCoefficients =
+      OldVectors.dense(
+        Array.fill($(factorSize) * numFeatures)(Random.nextGaussian() * $(initStd)) ++
```
`Random.nextGaussian()`
What about exposing seed to end users, by extending `HasSeed`?

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/regression/FMRegressor.scala:
```diff
@@ -0,0 +1,796 @@
+ * Model produced by [[FMRegressor]].
+ */
+@Since("3.0.0")
+class FMRegressorModel private[regression] (
```
`FMRegressorModel` -> `FMRegressionModel`

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/classification/FMClassifier.scala:
```diff
@@ -0,0 +1,333 @@
+  @transient private lazy val gradient = parseLoss(
+    LogisticLoss, $(factorSize), $(fitBias), $(fitLinear), numFeatures)
+
+  override protected def predictRaw(features: Vector): Vector = {
```
I know it will impl the prediction function twice, but impl it in the .ml side will avoid
1, vector conversion from .ml to .mllib
2, having two model coefficients in memory, the FMModel is usually much larger than other linear models.

mobai-zhanjf committed:
solve change request

zhengruifeng commented:
> I still doubt whether existing testsuite is enough.

1, I suggest to add several testsuites to check whether FMClassifier and FMRegressor can learn the intercept correctly.
Like '// Test if we can correctly learn Y = 0.1 + 1.2X1 - 1.3X2 + 20X1X2' in https://github.com/apache/spark/pull/5591/files

For example, we generate Y=Y = 0.1 + 1.2 X1 - 1.3 X2 + 20 X1 X2, and we can check whether 20 * X1 * X2 can be learned by checking  dot(V1, V2) ~== 20. 

2, I suggest to add some testcases with fitBias==false or/and fitLinear==false

mob-ai commented:
> > I still doubt whether existing testsuite is enough.
> 
> 1, I suggest to add several testsuites to check whether FMClassifier and FMRegressor can learn the intercept correctly.
> Like '// Test if we can correctly learn Y = 0.1 + 1.2X1 - 1.3X2 + 20X1X2' in https://github.com/apache/spark/pull/5591/files
> 
> For example, we generate Y=Y = 0.1 + 1.2 X1 - 1.3 X2 + 20 X1 X2, and we can check whether 20 * X1 * X2 can be learned by checking dot(V1, V2) ~== 20.
> 
> 2, I suggest to add some testcases with fitBias==false or/and fitLinear==false

for FMRegressor:
I already add testcases to cover above situation. The `generateFactorInteractionInput` function would random generate model weights then generate X correspondingly, `FMRegressor` need fit the weights correctly (include fitBias==false or/and fitLinear==false case).

for FMClassifier:
It is difficult to generate a dataset that FMClassifier fit perfectly. I generate weights randomly, then put in sigmoid function, it can't get 0/1 perfectly. If I want to generate 0/1 perfectly, the rawPrediction will very huge/small (sigmoid close to 1/0), but FMClassifier will fit the dataset using other model weights (the logloss still close to 0). So I only check the logloss correctness in FMClassifier.

zhengruifeng commented on mllib/src/test/scala/org/apache/spark/ml/classification/FMClassifierSuite.scala:
```diff
@@ -0,0 +1,147 @@
+    val fmRes = fmModel.transform(smallBinaryDataset)
+    val fmLogLoss = logLoss(fmRes)
+
+    val lr = new LogisticRegression()
```
I guess we do not need to compare FM with LR, check the convergen is enough.
I think we can more tests:
1, it seems that solver 'GD' is not tested at all;
2, it seemes that only dense datasets are used in testsuites;
3, add some cases like  'test("logistic regression doesn't fit intercept when fitIntercept is off")' / 'test("setThreshold, getThreshold")' / 'test("thresholds prediction")' / 'test("prediction on single instance")' in LogisticRegressionSuite

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/regression/FMRegressor.scala:
```diff
@@ -0,0 +1,826 @@
+
+  override def toString: String = {
+    s"FMRegressionModel: " +
+      s"uid = ${super.toString}, numFeatures = $numFeatures, " +
```
nit: delete space near '=' to follow current style.
`uid=${super.toString}, numFeatures=$numFeatures, ...`

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/classification/FMClassifier.scala:
```diff
@@ -0,0 +1,332 @@
+
+  override def toString: String = {
+    s"FMClassificationModel: " +
+      s"uid = ${super.toString}, numClasses = $numClasses, numFeatures = $numFeatures, " +
```
nit, delete space near `=`

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/regression/FMRegressor.scala:
```diff
@@ -0,0 +1,826 @@
+  }
+}
+
+private[ml] class AdamWUpdater(weightSize: Int) extends Updater with Logging {
```
Doc about this new solver is needed somewhere.

mob-ai replied:
> this regparam is appled to all coefficients, right?
yes, RegParam will apply to all coefficients. (I can't reply in above, so I put it here, sorry.)

zhengruifeng reviewed (commented):
In general, it looks good to me.
Since it is an important feature, I want to know how @srowen thinks of this PR?

zhengruifeng commented on mllib/src/test/scala/org/apache/spark/ml/regression/FMRegressorSuite.scala:
```diff
@@ -0,0 +1,240 @@
+      .setFitBias(fitBias)
+      .setFitLinear(fitLinear)
+      .setInitStd(0.01)
+      .setMaxIter(300)
```
MaxIter==300 maybe too large? Can we use a small maxIter for test performance?

mob-ai commented on mllib/src/test/scala/org/apache/spark/ml/regression/FMRegressorSuite.scala:
```diff
@@ -0,0 +1,240 @@
+    val rnd = new Random(seed)
+    val coefficientsSize = factorSize * numFeatures +
+      (if (fitLinear) numFeatures else 0) + (if (fitBias) 1 else 0)
+    val coefficients = Array.fill(coefficientsSize)(rnd.nextDouble() - 0.5)
```
yes, it's same in all checkMSE cases. But the label will different. So I think these test cases work.

mobai-zhanjf committed:
solve change request

srowen commented on mllib/src/main/scala/org/apache/spark/ml/classification/FMClassifier.scala:
```diff
@@ -0,0 +1,332 @@
+ * @note Multiclass labels are not currently supported.
+ */
+@Since("3.0.0")
+class FMClassifier @Since("3.0.0") (
```
It's not a big deal, but if the class is "since 3.0.0" then implicitly everything else is too. I don't think we need the annotations everywhere, but, I guess a bunch of classes do this, so w/e.

srowen commented on mllib/src/main/scala/org/apache/spark/ml/regression/FMRegressor.scala:
```diff
@@ -0,0 +1,839 @@
+   * @group param
+   */
+  @Since("3.0.0")
+  final val fitBias: BooleanParam = new BooleanParam(this, "fitBias",
```
How about calling this intercept and using HasIntercept?

srowen commented on mllib/src/main/scala/org/apache/spark/ml/regression/FMRegressor.scala:
```diff
@@ -0,0 +1,839 @@
+    require(coefficientsSize == coefficients.size,
+      s"coefficients.size did not match the excepted size ${coefficientsSize}")
+
+    val bias = if (fitBias) coefficients(coefficients.size - 1) else 0.0
```
General question: I think a lot of this code is copied from other implementations. Since you've looked at it recently, are there any reasonable opportunities to factor out common code? or is it to bitty and varies a bit across impls anyway?

mob-ai replied:
It's a suggestion of zhengruifeng. GD method uses mllib.Vector, but model predict uses ml.Vector.
So he suggested to impl it in the .ml side to avoid:
1, vector conversion from .ml to .mllib
2, having two model coefficients in memory, the FMModel is usually much larger than other linear models.

srowen commented on mllib/src/main/scala/org/apache/spark/ml/regression/FMRegressor.scala:
```diff
@@ -0,0 +1,839 @@
+        rawPrediction += weights(vWeightsSize + index) * value
+      }
+    }
+    (0 until factorSize).foreach { f =>
```
Can this calculation be refactored? It's repeated at least once, above

mob-ai replied:
The main difference is GD uses mllib.Vector, but model predict ml.Vector. Do you have suggestions to avoid Vector conversion? Or I remove one calculation, then ignore Vector conversion?

mobai-zhanjf committed:
fitBias change to fitIntercept

mobai-zhanjf committed:
resolve change requests

srowen commented:
Looking reasonable to me. @zhengruifeng ?

zhengruifeng commented:
LGTM.  @mob-ai You can open another ticket for doc & examples, after this pr get merged.

mob-ai commented:
> LGTM. @mob-ai You can open another ticket for doc & examples, after this pr get merged.

ok, "another ticket" means "another PR"?

mob-ai commented:
> LGTM. @mob-ai You can open another ticket for doc & examples, after this pr get merged.

ok, "another ticket" means "another PR"?

srowen commented:
You can make another PR for the same JIRA, or really, just add it here.

mobai-zhanjf committed:
add fm examples and docs

mob-ai commented:
add examples and docs

zhengruifeng commented:
@mob-ai  There seems still some comments from @srowen  has not been resolved ?

mob-ai commented:
> @mob-ai There seems still some comments from @srowen has not been resolved ?

All comments are resolved...I think...

Should I click all "resolve conversation" button? I realized just now...

srowen commented:
https://github.com/apache/spark/pull/26124#issuecomment-565390130
I think it'd be worthwhile to at least include a section in the MLlib docs on this new component. It can happen here rather than separately, if you're going to do it.

mobai-zhanjf committed:
update fm docs

mob-ai commented:
I added a new section for FM to describe the background of FM. And FM examples are placed in classification section and regression section.

srowen commented on docs/ml-classification-regression.md:
```diff
@@ -1013,6 +1050,44 @@ Refer to the [`IsotonicRegression` R API docs](api/R/spark.isoreg.html) for more
+
+Factorization machines are able to estimate interactions between features even in problems with huge sparsity.
+For more background and more details about the implementation of factorization machines,
+refer to [Factorization Machines section](ml-classification-regression.html#factorization-machines).
```
to -> to the

srowen commented on docs/ml-classification-regression.md:
```diff
@@ -1013,6 +1050,44 @@ Refer to the [`IsotonicRegression` R API docs](api/R/spark.isoreg.html) for more
+
+The following examples load a dataset in LibSVM format, split it into training and test sets,
+train on the first dataset, and then evaluate on the held-out test set.
+We scale features between 0 and 1 to prevent the exploding gradient problem.
```
between -> to be between

srowen commented on docs/ml-classification-regression.md:
```diff
@@ -1042,6 +1117,40 @@ regression](http://en.wikipedia.org/wiki/Tikhonov_regularization) model.
+# Factorization Machines
+
+[Factorization Machines](https://www.csie.ntu.edu.tw/~b97053/paper/Rendle2010FM.pdf) are able to estimate interactions
+between features even in problems with huge sparsity (like advertising and recommendation system).
```
You maybe don't have to repeat this sentence 3 times. You can omit it above?

srowen commented on docs/ml-classification-regression.md:
```diff
@@ -1042,6 +1117,40 @@ regression](http://en.wikipedia.org/wiki/Tikhonov_regularization) model.
+  \sum\limits^n_{i=1} \sum\limits^n_{j=i+1} \langle v_i, v_j \rangle x_i x_j
+$$
+
+First two terms denote intercept and linear term (as same as linear regression),
```
First -> The first
as same as -> same as in

srowen commented on docs/ml-classification-regression.md:
```diff
@@ -1042,6 +1117,40 @@ regression](http://en.wikipedia.org/wiki/Tikhonov_regularization) model.
+$$
+
+First two terms denote intercept and linear term (as same as linear regression),
+and last term denotes pairwise interactions term. $$v_i$$ describes the i-th variable
```
last -> the last

srowen commented on docs/ml-classification-regression.md:
```diff
@@ -1042,6 +1117,40 @@ regression](http://en.wikipedia.org/wiki/Tikhonov_regularization) model.
+and last term denotes pairwise interactions term. $$v_i$$ describes the i-th variable
+with k factors.
+
+FM can be used directly as the regression and optimization criterion is mean square error. FM also can be used as
```
"used directly for regression because the optimization criterion is mean squared error"?

srowen commented on docs/ml-classification-regression.md:
```diff
@@ -1042,6 +1117,40 @@ regression](http://en.wikipedia.org/wiki/Tikhonov_regularization) model.
+with k factors.
+
+FM can be used directly as the regression and optimization criterion is mean square error. FM also can be used as
+the binary classification through wrap sigmoid function, the optimization criterion is logit loss.
```
the binary -> for binary
What is a wrap sigmoid function.
, the optimization -> . The optimization criterion
Is it "logistic loss" BTW?

srowen commented on docs/ml-classification-regression.md:
```diff
@@ -1042,6 +1117,40 @@ regression](http://en.wikipedia.org/wiki/Tikhonov_regularization) model.
+This equation has only linear complexity in both k and n - i.e. its computation is in $$O(kn)$$.
+
+In general, in order to prevent the exploding gradient problem, it is best to scale continuous features between 0 and 1,
+or bin the continuous features and one-hot.
```
one-hot encode them

srowen reviewed (commented):
Looking pretty good

mobai-zhanjf committed:
update docs

mob-ai commented:
update fm docs

srowen commented:
Merged to master

srowen commented:
Ack, wait, this didn't pass tests after the last commit. Let me monitor

cloud-fan commented:
all the PR builders are broken now, with pyspark ML test failures. I'm reverting it to unblock other PRs. Please re-open it, thanks!

HyukjinKwon commented:
Thanks, all!

HeartSaVioR commented:
Ah OK that's the reason of pyspark failure. Nice finding and thanks for the quick action!

mob-ai commented:
@srowen I used following command to run pyspark tests. And I fixed FM python doc error.

```
python/run-tests --modules pyspark-ml
```

I'm very sorry about this.

I added a new commit, but not display in this PR. What should I do?

HeartSaVioR commented:
@mob-ai 
No worries. You can create a new PR with same content (title/description/commits) + additional commit.

mob-ai commented:
> @mob-ai
> No worries. You can create a new PR with same content (title/description/commits) + additional commit.

thanks!

srowen commented:
@cloud-fan yep this was entirely my mistake - looked at the wrong PR tab open in my browser to see if it had passed and merged it when it had not. My bad, I meant to go check and revert if needed this morning but you beat me to it
